### PR TITLE
implement pipe and pipe2

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/miri/issues/3746

### CODE REVIEW

File: src/shims/unix/unnamed_socket.rs [hunk 1]
```diff
@@ -55,17 +60,25 @@ impl FileDescription for SocketPair {
...
-            let available_space = MAX_SOCKETPAIR_BUFFER_CAPACITY.strict_sub(data_size);
-            if available_space != 0 {
+            if let Some(writebuf) = &peer_fd.downcast::<SocketPair>().unwrap().readbuf {
+                let data_size = writebuf.borrow().buf.len();
+                let available_space = MAX_SOCKETPAIR_BUFFER_CAPACITY.strict_sub(data_size);
+                if available_space != 0 {
+                    epoll_ready_events.epollout = true;
+                }
+            } else {
+                // Without a write buffer, writing never blocks.
```
Comment (oli-obk) [inline src/shims/unix/unnamed_socket.rs:81]: How can we have a peer without a read buffer?

Comment (RalfJung) [reply src/shims/unix/unnamed_socket.rs:81]: For the read end of a pipe, the peer is the write end of the pipe, and the write end doesn't have a read buffer.

Comment (oli-obk) [reply src/shims/unix/unnamed_socket.rs:81]: oh... right, that will just error

**Pull request merged**